File: 02-dynamic-programming/05_word_break.c
#include "../test_runner.h"

bool word_break(char* s, char** dict, int dict_size) {
    return false;
}

int main() {
    char* d1[] = {"leet","code"};
    TEST_BOOL("leetcode",   word_break("leetcode", d1, 2), true);
    char* d2[] = {"apple","pen"};
    TEST_BOOL("applepenapple", word_break("applepenapple", d2, 2), true);
    char* d3[] = {"cats","dog","sand","and","cat"};
    TEST_BOOL("catsandog",  word_break("catsandog", d3, 5), false);
    TEST_SUMMARY();
}

File: 02-dynamic-programming/05_word_break.py
import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests


# Word Break (LC #139) -- Medium
# Can string s be segmented into words from dictionary?
#
#   State: `dp[i]` = can s[0..i] be segmented?
#   Transition: `dp[i] = any(dp[j] and s[j:i] in dict)` for j < i

def word_break(s, word_dict):
    pass


run_tests(word_break, [
    (("leetcode", ["leet","code"]),             True),
    (("applepenapple", ["apple","pen"]),        True),
    (("catsandog", ["cats","dog","sand","and","cat"]), False),
    (("a", ["a"]),                              True),
    (("ab", ["a"]),                             False),
])
